execution/executor/src/components/apply_chunk_output.rs

```diff
@@ -153,124 +178,35 @@ impl ApplyChunkOutput {
-    ) -> Vec<Vec<(HashValue, &AccountStateBlob)>> {
-        account_blobs
-            .iter()
-            .map(|m| {
-                m.iter()
-                    .map(|(account, blob)| (HashAccountAddress::hash(account), blob))
-                    .collect::<Vec<_>>()
-            })
-            .collect()
-    }
-
-    fn parse_validator_set(accounts: &HashMap<AccountAddress, AccountState>) -> Result<EpochState> {
-        let validator_set = accounts
-            .get(&on_chain_config::config_address())
-            .map(|state| {
-                state
-                    .get_validator_set()?
-                    .ok_or_else(|| anyhow!("ValidatorSet does not exist"))
-            })
-            .ok_or_else(|| anyhow!("ValidatorSet account does not exist"))??;
-        let configuration = accounts
-            .get(&on_chain_config::config_address())
-            .map(|state| {
-                state
-                    .get_configuration_resource()?
-                    .ok_or_else(|| anyhow!("Configuration does not exist"))
-            })
-            .ok_or_else(|| anyhow!("Association account does not exist"))??;
-
-        Ok(EpochState {
-            epoch: configuration.epoch(),
-            verifier: (&validator_set).into(),
-        })
-    }
-
     fn assemble_ledger_diff(
-        to_keep: Vec<(Transaction, TransactionOutput)>,
+        to_keep: Vec<(Transaction, ParsedTransactionOutput)>,
         account_blobs: Vec<HashMap<AccountAddress, AccountStateBlob>>,
-        roots_with_node_hashes: Vec<(HashValue, HashMap<NibblePath, HashValue>)>,
+        new_node_hashes_vec: Vec<HashMap<NibblePath, HashValue>>,
+        state_checkpoint_hashes: Vec<Option<HashValue>>,
     ) -> (Vec<(Transaction, TransactionData)>, Vec<HashValue>) {
         let mut to_commit = vec![];
         let mut txn_info_hashes = vec![];
-        for ((txn, txn_output), ((state_tree_hash, new_node_hashes), blobs)) in itertools::zip_eq(
-            to_keep,
-            itertools::zip_eq(roots_with_node_hashes, account_blobs),
-        ) {
+        for (((txn, txn_output), state_checkpoint_hash), (new_node_hashes, blobs)) in
+            itertools::zip_eq(
+                itertools::zip_eq(to_keep, state_checkpoint_hashes),
+                itertools::zip_eq(new_node_hashes_vec, account_blobs),
+            )
+        {
+            let (write_set, events, reconfig_events, gas_used, status) = txn_output.unpack();
             let event_tree = {
-                let event_hashes: Vec<_> =
-                    txn_output.events().iter().map(CryptoHash::hash).collect();
+                let event_hashes: Vec<_> = events.iter().map(CryptoHash::hash).collect();
                 InMemoryAccumulator::<EventAccumulatorHasher>::from_leaves(&event_hashes)
             };
 
-            let txn_info = match txn_output.status() {
+            let state_change_hash = CryptoHash::hash(&write_set);
```

Comment:
pointer

Reply:
ah, that's true, each state_change_hash is an accumulation on previous state changes, so there's no need to compute a per block state change hash.